cpp/src/arrow/compute/kernels/vector_rolling.cc

```diff
@@ -0,0 +1,785 @@
+
+  void Append(ArgValue value, Status* st) {
+    sum = AddChecked::Call<ArgValue>(ctx, sum, value, st);
+  }
+  void Remove(ArgValue value, Status* st) {
+    sum = SubtractChecked::Call<ArgValue>(ctx, sum, value, st);
+  }
+  OutValue GetValue(Status* st) const { return sum; }
+};
+
+template <typename ArgType>
+struct ProdWindow {
+  using OutType = ArgType;
+  using ArgValue = typename GetViewType<ArgType>::T;
+  using OutValue = typename GetOutputType<OutType>::T;
+  KernelContext* ctx;
+  ArgValue prod = 1;
+  explicit ProdWindow(KernelContext* ctx) : ctx(ctx) {}
+
+  void Append(ArgValue value, Status* st) {
+    prod = Multiply::Call<ArgValue, ArgValue, ArgValue>(ctx, prod, value, st);
+  }
+  void Remove(ArgValue value, Status* st) {
+    prod = Divide::Call<ArgValue, ArgValue, ArgValue>(ctx, prod, value, st);
+  }
+  OutValue GetValue(Status* st) const { return prod; }
+};
+
+template <typename ArgType>
+struct ProdCheckedWindow {
+  using OutType = ArgType;
+  using ArgValue = typename GetViewType<ArgType>::T;
+  using OutValue = typename GetOutputType<OutType>::T;
+  KernelContext* ctx;
+  ArgValue prod = 1;
+  explicit ProdCheckedWindow(KernelContext* ctx) : ctx(ctx) {}
+
+  void Append(ArgValue value, Status* st) {
+    prod = MultiplyChecked::Call<ArgValue, ArgValue, ArgValue>(ctx, prod, value, st);
+  }
+  void Remove(ArgValue value, Status* st) {
+    prod = DivideChecked::Call<ArgValue, ArgValue, ArgValue>(ctx, prod, value, st);
+  }
+  OutValue GetValue(Status* st) const { return prod; }
+};
+
+template <typename ArgType>
+struct MinWindow {
+  using OutType = ArgType;
+  using ArgValue = typename GetViewType<ArgType>::T;
+  using OutValue = typename GetOutputType<OutType>::T;
+  KernelContext* ctx;
+  std::multiset<ArgValue> values;
+  explicit MinWindow(KernelContext* ctx) : ctx(ctx) {}
+  void Append(ArgValue value, Status* st) { values.insert(value); }
+  void Remove(ArgValue value, Status* st) {
+    auto it = values.find(value);
+    DCHECK_NE(it, values.end());
+    values.erase(it);
+  }
+  OutValue GetValue(Status* st) const { return *values.begin(); }
+};
+
+template <typename ArgType>
+struct MaxWindow {
+  using OutType = ArgType;
+  using ArgValue = typename GetViewType<ArgType>::T;
+  using OutValue = typename GetOutputType<OutType>::T;
+  KernelContext* ctx;
+  std::multiset<ArgValue, std::greater<ArgValue>> values;
+  explicit MaxWindow(KernelContext* ctx) : ctx(ctx) {}
+  void Append(ArgValue value, Status* st) { values.insert(value); }
+  void Remove(ArgValue value, Status* st) {
+    auto it = values.find(value);
+    DCHECK_NE(it, values.end());
+    values.erase(it);
+  }
+  OutValue GetValue(Status* st) const { return *values.begin(); }
+};
+
+template <typename ArgType>
+struct MeanWindow {
+  using OutType = DoubleType;
+  using ArgValue = typename GetViewType<ArgType>::T;
+  using OutValue = typename GetOutputType<OutType>::T;
+  KernelContext* ctx;
+  double sum = 0;
```

Comment:
Why `double` instead of `ArgValue` here? To avoid having to cast `sum` to `double` in `GetValue`? But how does that trade off with the error accumulation due to sums and subs on the `double`?

Reply:
This is to avoid overflow. Errors would only occur with large values, which are also susceptible to overflow errors. I think overflow is the worse problem here. Also, to improve accuracy, we can later switch to Kahan Sum or Pairwise Sum.